validator/coordinator/src/main/java/tech/pegasys/artemis/validator/coordinator/DepositProvider.java

```diff
@@ -0,0 +1,116 @@
+import tech.pegasys.artemis.datastructures.operations.DepositWithIndex;
+import tech.pegasys.artemis.datastructures.state.BeaconState;
+import tech.pegasys.artemis.datastructures.util.DepositUtil;
+import tech.pegasys.artemis.datastructures.util.MerkleTree;
+import tech.pegasys.artemis.datastructures.util.OptimizedMerkleTree;
+import tech.pegasys.artemis.pow.api.DepositEventChannel;
+import tech.pegasys.artemis.pow.event.DepositsFromBlockEvent;
+import tech.pegasys.artemis.storage.ChainStorageClient;
+import tech.pegasys.artemis.storage.api.FinalizedCheckpointEventChannel;
+import tech.pegasys.artemis.storage.events.FinalizedCheckpointEvent;
+import tech.pegasys.artemis.util.SSZTypes.SSZList;
+
+public class DepositProvider implements DepositEventChannel, FinalizedCheckpointEventChannel {
+
+  private final ChainStorageClient chainStorageClient;
+  private final MerkleTree depositMerkleTree = new OptimizedMerkleTree(DEPOSIT_CONTRACT_TREE_DEPTH);
+
+  private NavigableMap<UnsignedLong, DepositWithIndex> depositNavigableMap = new TreeMap<>();
+
+  public DepositProvider(ChainStorageClient chainStorageClient) {
+    this.chainStorageClient = chainStorageClient;
+  }
+
+  @Override
+  public void onDepositsFromBlock(DepositsFromBlockEvent event) {
+    event.getDeposits().stream()
+        .map(DepositUtil::convertDepositEventToOperationDeposit)
+        .forEach(
+            deposit -> {
+              synchronized (DepositProvider.this) {
+                depositNavigableMap.put(deposit.getIndex(), deposit);
+                depositMerkleTree.add(deposit.getData().hash_tree_root());
+              }
+            });
+  }
+
+  @Override
+  public synchronized void onFinalizedCheckpoint(FinalizedCheckpointEvent event) {
+    BeaconState finalizedState =
+        chainStorageClient
+            .getBlockState(event.getCheckpoint().getRoot())
+            .orElseThrow(
+                () -> new IllegalArgumentException("Finalized Checkpoint state can not be found."));
+
+    depositNavigableMap.headMap(finalizedState.getEth1_deposit_index()).clear();
+  }
+
+  public SSZList<Deposit> getDeposits(BeaconState state) {
+    UnsignedLong eth1DepositCount = state.getEth1_data().getDeposit_count();
+
+    UnsignedLong fromDepositIndex = state.getEth1_deposit_index();
+    UnsignedLong latestDepositIndexWithMaxBlock =
+        fromDepositIndex.plus(UnsignedLong.valueOf(MAX_DEPOSITS));
+
+    UnsignedLong toDepositIndex =
+        latestDepositIndexWithMaxBlock.compareTo(eth1DepositCount) > 0
+            ? eth1DepositCount
+            : latestDepositIndexWithMaxBlock;
+
+    return new SSZList<>(
+        getDepositsWithProof(fromDepositIndex, toDepositIndex, eth1DepositCount),
+        MAX_DEPOSITS,
+        Deposit.class);
+  }
+
+  public int getDepositMapSize() {
+    return depositNavigableMap.size();
+  }
+
+  // TODO: switch the MerkleTree to use UnsignedLongs instead of using toIntExact() here,
+  //  it will result in an overflow at some point
```

Comment:
Can we make sure we have a ticket for this please? Will take a while to get that many deposits but definitely don't want to be caught by surprise.

Reply:
https://pegasys1.atlassian.net/browse/BC-269